src/analysis/mesh/qgsmeshtriangulation.cpp

```diff
@@ -0,0 +1,316 @@
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmeshtriangulation.h"
+#include "qgsdualedgetriangulation.h"
+#include "qgsvectorlayer.h"
+#include "qgscoordinatetransformcontext.h"
+#include "qgscurve.h"
+#include "qgscurvepolygon.h"
+#include "qgsmultisurface.h"
+#include "qgsmulticurve.h"
+#include "qgsfeedback.h"
+
+QgsMeshTriangulation::QgsMeshTriangulation(): QObject()
+{
+  mTriangulation.reset( new QgsDualEdgeTriangulation() );
+}
+
+
+QgsMeshTriangulation::~QgsMeshTriangulation() = default;
+
+
+bool QgsMeshTriangulation::addVertices( QgsVectorLayer *vectorLayer,
+                                        int valueAttribute,
+                                        const QgsCoordinateTransformContext &transformContext,
+                                        QgsFeedback *feedback )
+{
+  if ( !vectorLayer )
+    return false;
+
+  if ( !vectorLayer->isValid() )
+    return false;
+
+  if ( feedback )
+    feedback->setProgress( 0 );
+
+  QgsCoordinateTransform transform( vectorLayer->crs(), mCrs, transformContext );
+
+  QgsFeatureIterator fIt = vectorLayer->getFeatures();
+  QgsFeature feat;
+  bool isZValueGeom = valueAttribute < 0;
+  int i = 0;
+  int featureCount = vectorLayer->featureCount();
+  while ( fIt.nextFeature( feat ) )
+  {
+    if ( feedback )
+    {
+      if ( feedback->isCanceled() )
+        break;
+
+      feedback->setProgress( 100 * i / featureCount );
+      i++;
+    }
+
+    QgsGeometry geom = feat.geometry();
+    geom.transform( transform, QgsCoordinateTransform::ForwardTransform, true );
+    QgsAbstractGeometry::vertex_iterator vit = geom.vertices_begin();
+
+    double value = 0;
+    if ( !isZValueGeom )
+      value = feat.attribute( valueAttribute ).toDouble();
+
+    while ( vit != geom.vertices_end() )
+    {
+      if ( feedback && feedback->isCanceled() )
+        break;
+      if ( isZValueGeom )
+        mTriangulation->addPoint( *vit );
+      else
+      {
+        ;
+        mTriangulation->addPoint( QgsPoint( QgsWkbTypes::PointZ, ( *vit ).x(), ( *vit ).y(), value ) );
+      }
+      ++vit;
+    }
+  }
+
+  return true;
+}
+
+bool QgsMeshTriangulation::addBreakLines( QgsVectorLayer *vectorLayer, int valueAttribute, const QgsCoordinateTransformContext &transformContext, QgsFeedback *feedback )
```

Comment:
Similar here -- better to use a QgsFeatureSource or QgsFeatureIterator instead of a raw vector layer